File: apps/ios/GalaxySSI/AgentExecutionLoopTimeline.swift
import Foundation

enum AgentExecutionLoopTimelineLabel: String, Codable, CaseIterable, Identifiable {
  case plan = "PLAN"
  case act = "ACT"
  case observe = "OBSERVE"
  case replan = "REPLAN"
  case verify = "VERIFY"
  case finalize = "FINALIZE"
  case learn = "LEARN"
  case waitingConfirmation = "WAITING_CONFIRMATION"
  case waitingResponse = "WAITING_RESPONSE"
  case paused = "PAUSED"
  case blocked = "BLOCKED"
  case failed = "FAILED"
  case cancelled = "CANCELLED"

  var id: String { rawValue }

  init?(phase: AgentExecutionLoopPhase) {
    guard phase != .completed else {
      return nil
    }
    self.init(rawValue: phase.rawValue)
  }
}

enum AgentExecutionLoopTimelineAction: String, Codable, CaseIterable, Identifiable {
  case pause = "PAUSE"
  case resume = "RESUME"
  case retry = "RETRY"
  case replan = "REPLAN"
  case cancel = "CANCEL"

  var id: String { rawValue }
}

struct AgentExecutionLoopTimelineProjection: Codable, Equatable {
  var controlEventType: AgentRunControlEventType
  var label: AgentExecutionLoopTimelineLabel?
  var stepId: String
  var toolCallId: String
  var payload: AgentRunControlPayload

  enum CodingKeys: String, CodingKey {
    case controlEventType = "control_event_type"
    case label
    case stepId = "step_id"
    case toolCallId = "tool_call_id"
    case payload
  }
}

enum AgentRunTimelineKind: String, Codable, CaseIterable, Identifiable {
  case plan = "PLAN"
  case tool = "TOOL"
  case result = "RESULT"
  case failure = "FAILURE"
  case retry = "RETRY"
  case act = "ACT"
  case observe = "OBSERVE"
  case verify = "VERIFY"
  case learn = "LEARN"
  case other = "OTHER"

  var id: String { rawValue }

  var payloadValue: String {
    rawValue.lowercased()
  }

  static func fromPayload(_ value: String?) -> AgentRunTimelineKind? {
    let normalized = value?.trimmingCharacters(in: .whitespacesAndNewlines).uppercased() ?? ""
    return allCases.first { $0.rawValue == normalized }
  }
}

struct AgentRunTimelineCoverage: Codable, Equatable {
  var hasPlan: Bool
  var toolEventCount: Int
  var hasResult: Bool
  var hasFailure: Bool
  var retryEventCount: Int

  var terminal: Bool {
    hasResult || hasFailure
  }

  var complete: Bool {
    hasPlan && terminal
  }

  enum CodingKeys: String, CodingKey {
    case hasPlan = "has_plan"
    case toolEventCount = "tool_event_count"
    case hasResult = "has_result"
    case hasFailure = "has_failure"
    case retryEventCount = "retry_event_count"
  }
}

enum AgentRunTimelineContract {
  static let version = "galaxyssi.run-timeline/1.0"

  static func kind(_ event: AgentRunControlEvent) -> AgentRunTimelineKind {
    if let declared = AgentRunTimelineKind.fromPayload(event.payload["timeline_kind"]?.stringValue) {
      return declared
    }
    switch event.type {
    case .planning:
      return .plan
    case .toolStarted, .toolProgress, .toolCompleted, .toolPermissionRequired:
      return .tool
    case .retrying, .runRecovered:
      return .retry
    case .runCompleted:
      return .result
    case .runFailed, .runCancelled:
      return .failure
    default:
      return .other
    }
  }

  static func coverage(_ events: [AgentRunControlEvent]) -> AgentRunTimelineCoverage {
    let kinds = events.map(kind)
    return AgentRunTimelineCoverage(
      hasPlan: kinds.contains(.plan),
      toolEventCount: kinds.filter { $0 == .tool }.count,
      hasResult: kinds.contains(.result),
      hasFailure: kinds.contains(.failure),
      retryEventCount: kinds.filter { $0 == .retry }.count
    )
  }
}

enum AgentExecutionLoopTimelinePolicy {
  static func actionsForPhase(_ phase: AgentPhase) -> [AgentExecutionLoopTimelineAction] {
    switch phase {
    case .planning, .waitingConfirmation, .executing, .verifying:
      return [.pause, .cancel]
    case .observing, .waitingResponse:
      return [.cancel]
    case .paused:
      return [.resume, .cancel]
    case .blocked:
      return [.replan, .cancel]
    case .failed:
      return [.retry, .replan]
    case .cancelled, .completed:
      return []
    }
  }

  static func project(_ event: AgentExecutionLoopEvent) -> AgentExecutionLoopTimelineProjection {
    let recovered = event.previousPhase.map { [.blocked, .failed].contains($0) } == true && event.phase.isActive
    let phaseType: AgentRunControlEventType
    switch event.phase {
    case .plan:
      phaseType = .planning
    case .act:
      phaseType = event.toolCall ? .toolStarted : .stepStarted
    case .observe:
      phaseType = .toolProgress
    case .replan:
      phaseType = .retrying
    case .verify:
      phaseType = .toolProgress
    case .finalize, .learn:
      phaseType = .stepCompleted
    case .waitingConfirmation:
      phaseType = .waitingForUser
    case .waitingResponse:
      phaseType = .waitingForDevice
    case .paused:
      phaseType = .paused
    case .blocked, .failed:
      phaseType = .runFailed
    case .cancelled:
      phaseType = .runCancelled
    case .completed:
      phaseType = .runCompleted
    }
    let timelineKind: AgentRunTimelineKind
    switch event.phase {
    case .plan:
      timelineKind = .plan
    case .act:
      timelineKind = event.toolCall ? .tool : .act
    case .observe:
      timelineKind = .observe
    case .replan:
      timelineKind = .retry
    case .verify:
      timelineKind = .verify
    case .finalize, .completed:
      timelineKind = .result
    case .learn:
      timelineKind = .learn
    case .blocked, .failed, .cancelled:
      timelineKind = .failure
    case .waitingConfirmation, .waitingResponse, .paused:
      timelineKind = .other
    }
    let actionId = event.snapshot.lastActionId
    return AgentExecutionLoopTimelineProjection(
      controlEventType: recovered ? .runRecovered : phaseType,
      label: AgentExecutionLoopTimelineLabel(phase: event.phase),
      stepId: actionId,
      toolCallId: event.toolCall ? actionId : "",
      payload: [
        "timeline_contract": .string(AgentRunTimelineContract.version),
        "timeline_kind": .string(timelineKind.payloadValue),
        "loop_phase": .string(event.phase.rawValue.lowercased()),
        "previous_loop_phase": .string(event.previousPhase?.rawValue.lowercased() ?? ""),
        "loop_revision": .int(event.snapshot.revision),
        "loop_reason": .string(event.reason),
        "loop_task_id": .string(event.snapshot.taskId),
        "loop_action_id": .string(actionId),
        "loop_retry": .bool(event.retry),
        "loop_tool_call": .bool(event.toolCall),
        "loop_iterations": .int(Int64(event.snapshot.usage.iterations)),
        "loop_actions": .int(Int64(event.snapshot.usage.actions)),
        "loop_replans": .int(Int64(event.snapshot.usage.replans)),
        "loop_tool_calls": .int(Int64(event.snapshot.usage.toolCalls)),
        "loop_retries": .int(Int64(event.snapshot.usage.retries)),
        "loop_active_ms": .int(event.snapshot.usage.activeDurationMillis),
        "loop_budget_failure": .string(event.snapshot.budgetFailure)
      ]
    )
  }

  static func isSameRevision(event: AgentRunControlEvent?, revision: Int64) -> Bool {
    event?.payload["loop_revision"]?.intValue == revision
  }

  static func transcriptDedupeKey(turnId: String, event: AgentExecutionLoopEvent) -> String {
    "agent-loop:\(turnId):\(event.phase.rawValue):\(event.snapshot.revision)"
  }

  static func phaseFromTranscriptDedupeKey(_ value: String) -> AgentExecutionLoopPhase? {
    guard value.hasPrefix("agent-loop:") else {
      return nil
    }
    let parts = value.split(separator: ":").map(String.init)
    guard parts.count > 2 else {
      return nil
    }
    let phaseName = parts[2].trimmingCharacters(in: .whitespacesAndNewlines).uppercased()
    return AgentExecutionLoopPhase.allCases.first { $0.rawValue == phaseName }
  }

  static func suppressSupersededPlaceholders(_ entries: [AgentTranscriptEntry]) -> [AgentTranscriptEntry] {
    let hasToolStart = entries.contains { $0.dedupeKey.contains(":TOOL_STARTED:") }
    let hasToolCompletion = entries.contains { $0.dedupeKey.contains(":TOOL_COMPLETED:") }
    return entries.filter { entry in
      let phase = phaseFromTranscriptDedupeKey(entry.dedupeKey)
      if phase == .act {
        return !hasToolStart
      }
      if phase == .observe {
        return !hasToolCompletion
      }
      return true
    }
  }
}

File: apps/android/app/src/main/java/com/galaxyssi/chat/AgentExecutionLoopTimeline.kt
package com.galaxyssi.chat

import java.util.Locale

enum class AgentExecutionLoopTimelineLabel {
    PLAN,
    ACT,
    OBSERVE,
    REPLAN,
    VERIFY,
    FINALIZE,
    LEARN,
    WAITING_CONFIRMATION,
    WAITING_RESPONSE,
    PAUSED,
    BLOCKED,
    FAILED,
    CANCELLED
}

enum class AgentExecutionLoopTimelineAction {
    PAUSE,
    RESUME,
    RETRY,
    REPLAN,
    CANCEL
}

data class AgentExecutionLoopTimelineProjection(
    val controlEventType: AgentRunControlEventType,
    val label: AgentExecutionLoopTimelineLabel?,
    val stepId: String,
    val toolCallId: String,
    val payload: AgentNativeJsonObject
)

enum class AgentRunTimelineKind {
    PLAN,
    TOOL,
    RESULT,
    FAILURE,
    RETRY,
    ACT,
    OBSERVE,
    VERIFY,
    LEARN,
    OTHER
}

data class AgentRunTimelineCoverage(
    val hasPlan: Boolean,
    val toolEventCount: Int,
    val hasResult: Boolean,
    val hasFailure: Boolean,
    val retryEventCount: Int
) {
    val terminal: Boolean
        get() = hasResult || hasFailure
    val complete: Boolean
        get() = hasPlan && terminal
}

object AgentRunTimelineContract {
    const val VERSION = "galaxyssi.run-timeline/1.0"

    fun kind(event: AgentRunControlEvent): AgentRunTimelineKind {
        val declared = event.payload["timeline_kind"]
            ?.toString()
            ?.trim()
            ?.uppercase(Locale.ROOT)
            ?.let { value -> AgentRunTimelineKind.entries.firstOrNull { it.name == value } }
        if (declared != null) return declared
        return when (event.type) {
            AgentRunControlEventType.PLANNING -> AgentRunTimelineKind.PLAN
            AgentRunControlEventType.TOOL_STARTED,
            AgentRunControlEventType.TOOL_PROGRESS,
            AgentRunControlEventType.TOOL_COMPLETED,
            AgentRunControlEventType.TOOL_PERMISSION_REQUIRED -> AgentRunTimelineKind.TOOL
            AgentRunControlEventType.RETRYING,
            AgentRunControlEventType.RUN_RECOVERED -> AgentRunTimelineKind.RETRY
            AgentRunControlEventType.RUN_COMPLETED -> AgentRunTimelineKind.RESULT
            AgentRunControlEventType.RUN_FAILED,
            AgentRunControlEventType.RUN_CANCELLED -> AgentRunTimelineKind.FAILURE
            else -> AgentRunTimelineKind.OTHER
        }
    }

    fun coverage(events: List<AgentRunControlEvent>): AgentRunTimelineCoverage {
        val kinds = events.map(::kind)
        return AgentRunTimelineCoverage(
            hasPlan = AgentRunTimelineKind.PLAN in kinds,
            toolEventCount = kinds.count { it == AgentRunTimelineKind.TOOL },
            hasResult = AgentRunTimelineKind.RESULT in kinds,
            hasFailure = AgentRunTimelineKind.FAILURE in kinds,
            retryEventCount = kinds.count { it == AgentRunTimelineKind.RETRY }
        )
    }
}

object AgentExecutionLoopTimelinePolicy {
    fun actionsForPhase(phase: AgentPhase): List<AgentExecutionLoopTimelineAction> = when (phase) {
        AgentPhase.PLANNING,
        AgentPhase.WAITING_CONFIRMATION,
        AgentPhase.EXECUTING,
        AgentPhase.VERIFYING -> listOf(
            AgentExecutionLoopTimelineAction.PAUSE,
            AgentExecutionLoopTimelineAction.CANCEL
        )
        AgentPhase.OBSERVING,
        AgentPhase.WAITING_RESPONSE -> listOf(AgentExecutionLoopTimelineAction.CANCEL)
        AgentPhase.PAUSED -> listOf(
            AgentExecutionLoopTimelineAction.RESUME,
            AgentExecutionLoopTimelineAction.CANCEL
        )
        AgentPhase.BLOCKED -> listOf(
            AgentExecutionLoopTimelineAction.REPLAN,
            AgentExecutionLoopTimelineAction.CANCEL
        )
        AgentPhase.FAILED -> listOf(
            AgentExecutionLoopTimelineAction.RETRY,
            AgentExecutionLoopTimelineAction.REPLAN
        )
        AgentPhase.CANCELLED,
        AgentPhase.COMPLETED -> emptyList()
    }

    fun project(event: AgentExecutionLoopEvent): AgentExecutionLoopTimelineProjection {
        val recovered = event.previousPhase in setOf(
            AgentExecutionLoopPhase.BLOCKED,
            AgentExecutionLoopPhase.FAILED
        ) && event.phase.isActive
        val phaseType = when (event.phase) {
            AgentExecutionLoopPhase.PLAN -> AgentRunControlEventType.PLANNING
            AgentExecutionLoopPhase.ACT -> if (event.toolCall) {
                AgentRunControlEventType.TOOL_STARTED
            } else {
                AgentRunControlEventType.STEP_STARTED
            }
            AgentExecutionLoopPhase.OBSERVE -> AgentRunControlEventType.TOOL_PROGRESS
            AgentExecutionLoopPhase.REPLAN -> AgentRunControlEventType.RETRYING
            AgentExecutionLoopPhase.VERIFY -> AgentRunControlEventType.TOOL_PROGRESS
            AgentExecutionLoopPhase.FINALIZE -> AgentRunControlEventType.STEP_COMPLETED
            AgentExecutionLoopPhase.LEARN -> AgentRunControlEventType.STEP_COMPLETED
            AgentExecutionLoopPhase.WAITING_CONFIRMATION -> AgentRunControlEventType.WAITING_FOR_USER
            AgentExecutionLoopPhase.WAITING_RESPONSE -> AgentRunControlEventType.WAITING_FOR_DEVICE
            AgentExecutionLoopPhase.PAUSED -> AgentRunControlEventType.PAUSED
            AgentExecutionLoopPhase.BLOCKED -> AgentRunControlEventType.RUN_FAILED
            AgentExecutionLoopPhase.FAILED -> AgentRunControlEventType.RUN_FAILED
            AgentExecutionLoopPhase.CANCELLED -> AgentRunControlEventType.RUN_CANCELLED
            AgentExecutionLoopPhase.COMPLETED -> AgentRunControlEventType.RUN_COMPLETED
        }
        val controlType = if (recovered) AgentRunControlEventType.RUN_RECOVERED else phaseType
        val label = event.phase.takeUnless { it == AgentExecutionLoopPhase.COMPLETED }
            ?.let { AgentExecutionLoopTimelineLabel.valueOf(it.name) }
        val actionId = event.snapshot.lastActionId
        val timelineKind = when (event.phase) {
            AgentExecutionLoopPhase.PLAN -> AgentRunTimelineKind.PLAN
            AgentExecutionLoopPhase.ACT -> if (event.toolCall) {
                AgentRunTimelineKind.TOOL
            } else {
                AgentRunTimelineKind.ACT
            }
            AgentExecutionLoopPhase.OBSERVE -> AgentRunTimelineKind.OBSERVE
            AgentExecutionLoopPhase.REPLAN -> AgentRunTimelineKind.RETRY
            AgentExecutionLoopPhase.VERIFY -> AgentRunTimelineKind.VERIFY
            AgentExecutionLoopPhase.FINALIZE,
            AgentExecutionLoopPhase.COMPLETED -> AgentRunTimelineKind.RESULT
            AgentExecutionLoopPhase.LEARN -> AgentRunTimelineKind.LEARN
            AgentExecutionLoopPhase.BLOCKED,
            AgentExecutionLoopPhase.FAILED,
            AgentExecutionLoopPhase.CANCELLED -> AgentRunTimelineKind.FAILURE
            AgentExecutionLoopPhase.WAITING_CONFIRMATION,
            AgentExecutionLoopPhase.WAITING_RESPONSE,
            AgentExecutionLoopPhase.PAUSED -> AgentRunTimelineKind.OTHER
        }
        return AgentExecutionLoopTimelineProjection(
            controlEventType = controlType,
            label = label,
            stepId = actionId,
            toolCallId = actionId.takeIf { event.toolCall }.orEmpty(),
            payload = mapOf(
                "timeline_contract" to AgentRunTimelineContract.VERSION,
                "timeline_kind" to timelineKind.name.lowercase(Locale.ROOT),
                "loop_phase" to event.phase.name.lowercase(Locale.ROOT),
                "previous_loop_phase" to event.previousPhase?.name.orEmpty().lowercase(Locale.ROOT),
                "loop_revision" to event.snapshot.revision,
                "loop_reason" to event.reason,
                "loop_task_id" to event.snapshot.taskId,
                "loop_action_id" to actionId,
                "loop_retry" to event.retry,
                "loop_tool_call" to event.toolCall,
                "loop_iterations" to event.snapshot.usage.iterations,
                "loop_actions" to event.snapshot.usage.actions,
                "loop_replans" to event.snapshot.usage.replans,
                "loop_tool_calls" to event.snapshot.usage.toolCalls,
                "loop_retries" to event.snapshot.usage.retries,
                "loop_active_ms" to event.snapshot.usage.activeDurationMillis,
                "loop_budget_failure" to event.snapshot.budgetFailure
            )
        )
    }

    fun isSameRevision(
        event: AgentRunControlEvent?,
        revision: Long
    ): Boolean = event?.payload?.get("loop_revision").toString().toLongOrNull() == revision

    fun transcriptDedupeKey(turnId: String, event: AgentExecutionLoopEvent): String =
        "agent-loop:$turnId:${event.phase.name}:${event.snapshot.revision}"

    fun phaseFromTranscriptDedupeKey(value: String): AgentExecutionLoopPhase? {
        if (!value.startsWith("agent-loop:")) return null
        val phaseName = value.split(':').getOrNull(2).orEmpty()
        return AgentExecutionLoopPhase.entries.firstOrNull { it.name == phaseName }
    }

    fun suppressSupersededPlaceholders(
        entries: List<AgentTranscriptEntry>
    ): List<AgentTranscriptEntry> {
        val hasToolStart = entries.any { it.dedupeKey.contains(":TOOL_STARTED:") }
        val hasToolCompletion = entries.any { it.dedupeKey.contains(":TOOL_COMPLETED:") }
        return entries.filterNot { entry ->
            when (phaseFromTranscriptDedupeKey(entry.dedupeKey)) {
                AgentExecutionLoopPhase.ACT -> hasToolStart
                AgentExecutionLoopPhase.OBSERVE -> hasToolCompletion
                else -> false
            }
        }
    }
}
